art/attacks/evasion/shadow_attack.py

```diff
@@ -0,0 +1,318 @@
+        )
+
+        for _ in range(self.nb_steps):
+
+            gradients_ce = np.mean(
+                self.estimator.loss_gradient(x=x_batch + perturbation, y=y_batch, sampling=False)
+                * (1 - 2 * int(self.targeted)),
+                axis=0,
+                keepdims=True,
+            )
+
+            gradients = gradients_ce - self._get_regularisation_loss_gradients(perturbation)
+
+            perturbation += self.learning_rate * gradients
+
+        x_p = x + perturbation
+
+        x_adv = np.clip(x_p, a_min=self.estimator.clip_values[0], a_max=self.estimator.clip_values[1]).astype(
+            ART_NUMPY_DTYPE
+        )
+
+        return x_adv
+
+    def _get_regularisation_loss_gradients(self, perturbation):
+        """
+        Get regularisation loss gradients.
+
+        :param perturbation: The perturbation to be regularised.
+        :type perturbation: `np.ndarray`
+        :return: The loss gradients of the perturbation.
+        """
+
+        if self.frame_work == "tensorflow":
+
+            import tensorflow as tf
+
+            if self.estimator.channel_index == 3:
+                perturbation = perturbation.transpose(0, 3, 1, 2)
+
+            if tf.executing_eagerly():
+                with tf.GradientTape() as tape:
+
+                    perturbation_t = tf.convert_to_tensor(perturbation)
+                    tape.watch(perturbation_t)
+
+                    x_ = perturbation_t[:, :, :, 1:] - perturbation_t[:, :, :, :-1]
+                    y_ = perturbation_t[:, :, 1:, :] - perturbation_t[:, :, :-1, :]
+
+                    loss_tv = tf.reduce_sum(x_ * x_, axis=(1, 2, 3)) + tf.reduce_sum(y_ * y_, axis=(1, 2, 3))
+
+                    if perturbation_t.shape[1] == 1:
+                        loss_s = 0.0
+                    elif perturbation_t.shape[1] == 3:
+                        loss_s = tf.norm(
+                            (perturbation_t[:, 0, :, :] - perturbation_t[:, 1, :, :]) ** 2
+                            + (perturbation_t[:, 1, :, :] - perturbation_t[:, 2, :, :]) ** 2
+                            + (perturbation_t[:, 0, :, :] - perturbation_t[:, 2, :, :]) ** 2,
+                            ord=2,
+                            axis=(1, 2),
+                        )
+
+                    loss_c = tf.norm(tf.reduce_mean(tf.abs(perturbation_t), axis=[2, 3]), ord=2, axis=1) ** 2
+
+                    loss = self.lambda_tv * loss_tv + self.lambda_s * loss_s + self.lambda_c * loss_c
+
+                    gradients = tape.gradient(loss, perturbation_t).numpy()
+
+                    if self.estimator.channel_index == 3:
+                        gradients = gradients.transpose(0, 2, 3, 1)
+
+            else:
+                raise ValueError("Expecting eager execution.")
+
+        elif self.frame_work == "pytorch":
+
+            import torch
+
+            perturbation_t = torch.from_numpy(perturbation).to("cpu")
```

Comment:
My thought was that the calculation of the regularisation loss will always be a small computation and it would never get any advantage from running on a GPU.